### DISCUSSION TITLE
Typing Invoke Services

### DISCUSSION THREAD

**Discussion by Gil3co** [statelyai/xstate#3534] [Q&A]
Hi, I'm a beginner in xstate, and after racking my brains and searching the web, I couldn't find a way to type invoke services.

For example, I would like to strongly type the following service:
` fetchRecipe: async (context: IContext, event: any) => {
          const { apolloClient, userId, curRecipe } = context;
          if (curRecipe._id !== dummyUserId) return curRecipe;
          const { recipeId } = event;
          const { data } = await apolloClient.query<IGetRecipe>({
            query: getRecipeQuery,
            variables: {
              recipeId,
              userId,
            },
          });
          if (data) {
            return data.getRecipe;
          }
        },`

I know that `event` should hold a type ("SEND") and a recipeId, which is a string.

Any way I could do that?

The only option that I've found is to type all of the events in the machine's schema:
`events: {} as {type: "SEND"; recipeId: string} | TAllOtherEvents`

But then, in order to make use of recipeId being a string, I would have to verify that `event.type === "SEND"`.

Any suggestions?

**Comment by Andarist** [comment]
It feels like you are looking for our [typegen](https://xstate.js.org/docs/guides/typescript.html#typegen). It should be able to figure out what kind of events u can receive when invoking `fetchRecipe`. Note that to make it work it's recommended to move `fetchRecipe` to the second argument of the `createMachine`.

The final code would look something like this:

```ts
import { createMachine } from "xstate";

createMachine(
  {
    context: {} as {
      curRecipe: Recipe,
      userId: string,
      apolloClient: ApolloClient,
    },
    schema: {
      events: {} as { type: "SEND"; recipeId: string } | TAllOtherEvents,
    },
    tsTypes: {} as import("./myMachine.typegen").Typegen0,
    initial: "a",
    states: {
      a: {
        on: {
          SEND: "b",
        },
      },
      b: {
        invoke: {
          src: "fetchRecipe",
        },
      },
    },
  },
  {
    fetchRecipe: async (context, event) => {
      const { apolloClient, userId, curRecipe } = context;
      if (curRecipe._id !== dummyUserId) return curRecipe;
      const { recipeId } = event;
      const { data } = await apolloClient.query<IGetRecipe>({
        query: getRecipeQuery,
        variables: { recipeId, userId },
      });
      if (data) {
        return data.getRecipe;
      }
    },
  }
);

```

**Reply by Gil3co** [reply]
Thanks, but it still doesn't work as it should, maybe I'm missing something:
```TypeScript
createMachine({
  schema: {
    context: {} as IContext,
    events: {} as { type: "GO_TO_RECIPE"; recipeId: string },
  },
  tsTypes: {} as import("./mainMachine.typegen").Typegen0,
  initial: "recipeView",
  states: {
    recipeView: {
      initial: "loading",
      states: {
        loading: {
          invoke: {
            id: "fetchRecipe",
            src: "fetchRecipe",
          },
        },
      },
    },
  },
  services: {
    fetchRecipe: async (context: IContext, event) => {
      const { apolloClient, userId, curRecipe } = context;
      if (curRecipe._id !== dummyUserId) return curRecipe;
      if (event.type === "GO_TO_RECIPE") {
       // event.recipeId is a string and doesn't show a TS error
      }
      const { recipeId } = event;
      const { data } = await apolloClient.query<IGetRecipe>({
        query: getRecipeQuery,
        variables: {
          recipeId,
          userId,
        },
      });
      if (data) {
        return data.getRecipe;
      }
    },
  },
});
```

Only if I check event.type === "GO_TO_RECIPE" will I be able to use the types.
In addition, if I comment `  tsTypes: {} as import("./mainMachine.typegen").Typegen0`, I get the same result and have to check the event.type.
Maybe I'm using typegen wrong?

**Reply by Andarist** [reply]
Could you share this in a runnable form? You could copy-paste this machine and your generated typegen into a codesandbox playground

**Reply by Gil3co** [reply]
https://codesandbox.io/s/elated-spence-1cjexx?file=/src/mainMachine.ts

It seems to work as the code is now, but, if you uncomment `DO_NOTHING`, you will see an error inside `fetchRecipe`

**Reply by Andarist** [reply]
We might see in the generated typegen this:
```ts
  eventsCausingServices: {
    fetchRecipe: "DO_NOTHING" | "GO_TO_RECIPE";
  };
```

That's why the `DO_NOTHING` event gets to this service. I can't be sure **but** I think that you didn't regenerate the typegen after changing the machine's config (we can notice a commented out transition for `DO_NOTHING` in the codesandbox).

**Reply by Gil3co** [reply]
I haven't regenerated tha typegen, on purpose, becuase I wanted to show you that it works with the typegen, but **only** when I comment the DO_NOTHING event.

I replaced the typegen with a regenerated version.

**Reply by Andarist** [reply]
As far as I can see - it works now. I've included `{ type: 'DO_NOTHING' }` in the `schema.events` and I don't have any error reported in the `fetchRecipe`:
https://codesandbox.io/s/hungry-marco-y3eo3j?file=/src/mainMachine.ts:903-908

**Reply by Gil3co** [reply]
This is really, really weird. I copied your code into my sandbox and it seems to work as well now. Thanks a lot, I will try to implement this with all the rest of my services

Thanks!
